src/Core/SearchEngine.vala

```diff
@@ -0,0 +1,63 @@
+/*
+ * SPDX-FileCopyrightText: 2024 elementary, Inc. (https://elementary.io)
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ *
+ * Authored by: Leonhard Kargl <leo.kargl@proton.me>
+ */
+
+public class AppCenterCore.SearchEngine : Object {
+    public ListModel results { get; construct; }
+
+    private ListStore packages;
+    private AppStream.Pool pool;
+
+    private string[] query;
+    private AppStream.Category? category;
+
+    public SearchEngine (Package[] packages, AppStream.Pool pool) {
+        var unique_packages = new Gee.HashMap<string, Package> ();
+        foreach (var package in packages) {
+            var package_component_id = package.normalized_component_id;
+            if (unique_packages.has_key (package_component_id)) {
+                if (package.origin_score > unique_packages[package_component_id].origin_score) {
+                    unique_packages[package_component_id] = package;
+                }
+            } else {
+                unique_packages[package_component_id] = package;
+            }
+        }
+
+        this.packages.splice (0, 0, unique_packages.values.to_array ());
+        this.pool = pool;
+    }
```

Comment:
Yeah I didn't use it here because we don't need the packages array but only the liststore with the unique packages so it seemed kinda pointless to first put the array into a property and the in construct extracting this stuff and then setting the property to null to free the memory. Alternatively we could ofc do all of this in the flatpackbackend and hand over the liststore right away :shrug:

Reply:
Yeah that might be the way to go here so we can stick to coding conventions